### ISSUE TITLE
4 Stem Separation Mode only works once with CPU as Inference Device

### ISSUE THREAD

**Issue by Rosenheim23** [intel/openvino-plugins-ai-audacity#1]
Hello,

I recently came across this openvino plugin for Audacity.

I am using it to separate the drums from music tracks and have encountered the following problem:

If I select the CPU as Inference Device in the 4 Stem method, the separation works once. If I then do the same process again, whether with the same piece of music or a different one, I only get noise in the 4 separated stems. Only when I delete the .blob files from the cache in the profile directory does the separation work again for one time.

If the GPU is selected as the inference device, the separation always works.

I have tested this on 2 notebooks, one with an 8th generation i5 CPU and Windows 10 and one with an 11th generation i5 CPU and Windows 11.

I use the CPU because in my case the GPU is significantly slower.

Many thanks and best regards
Christian

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23, thanks for reaching out. This is strange behavior... I tried to reproduce it on my 11th Gen i7, but it seems to produce good separation results on back-to-back runs (i.e. when picking up pre-compiled blobs from cache dir).

Couple follow up questions though.
1. Can you give exact CPU model used? For example, in my case I'm using i7-1185G7.
2. Just curious what the format (mono/stereo, sample rate) of the input track is, and also the selected length. Can you also reproduce the issue for 'short' selections (say, ~10 seconds or so)?

Thanks!
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Hmm, interesting. I highly doubt that using an ASIO-enabled custom build of Audacity would affect the behavior here... as it seems to be related to OpenVINO's caching mechanism. I have a very similar setup (tiger lake i7, Windows 11), and so I'm surprised we see different behavior.

Maybe to be 100% sure, it would be worth it to test on your side using the release builds of Audacity? But again, I'd be shocked to see any behavioral differences..

I'll also think on this a bit more, and try to come up with some way to half-split the problem..

Thanks,
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Sorry for the late reply. Any chance you can post a brief video showing the problem? I still am unable to reproduce this...

Thanks,
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

> Here is a .zip file of the recording:
Thank you for taking the time to do that. It's very helpful. 

Out of curiosity, given that it's a German version of Windows, are there any 'special' characters in %appdata% path? I had heard recently that there was an OpenVINO bug around certain kinds of special characters (such as chinese characters) in the model path -- maybe it could be something similar?

Do you encounter weird caching behavior for other effects? For example, if you ran whisper transcription on the vocal track for 'Under Pressure', do you get consistent results for back-to-back runs?

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Thanks for the quick reply. It's very useful to know that there seems to be some *overall* caching issue happening, as opposed to something specific to the music separation model. 

I've submitted a question to OpenVINO issues page here (https://github.com/openvinotoolkit/openvino/issues/21942) to understand if there are any known issues with model caching on OpenVINO 2023.1.

Thanks,
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Thanks Christian! Glad it's still usable despite the caching bug. 

At some point in the next week or so, I might compile the plugins against the latest OpenVINO release (2023.2), and post a new DLL package here to see if that resolves your issue (if you're willing to test). 

And I'll keep trying to reproduce this on my side (I just tried on my wife's laptop 😆 -- 7th gen i5 -- but didn't hit the issue there either).

Best Regards,
Ryan

**Comment by akash07k** [comment]
Hi @RyanMetcalfeInt8 
I too am impacted with issue, but on 2 Stem separation mode also.
I'll create a new issue for this though.

**Comment by RyanMetcalfeInt8** [comment]
> but on 2 Stem separation mode also.

Note that under the hood, 2-stem and 4-stem run the same model (the 4-stem model). The 2-stem just applies "mix-and-render" for the drums, bass, and 'other instruments' tracks.

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23, @akash07k,

Sorry for the delay in getting a test package built with latest OpenVINO (2023.2). Here it is, if you're still willing to try it out to see if it resolves the caching bug:

[Audacity-OpenVINO-2023.2-Module-3.4.2-R1](https://drive.google.com/file/d/1mPQakma-XTCdQUHykxUOMAt0ygYMnqo2/view?usp=sharing)

Install it in the same way as original Audacity-OpenVINO-Module-3.4.2-R1.zip, but allow it to overwrite any existing files. You'll need to set mod-openvino to 'Enabled' again after install (and close / re-open Audacity).

Let me know if you have any questions -- looking forward to hearing the results.

Thanks again!
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Thanks for the super quick turnaround on that! Just so I'm clear, which platform / configuration did you try it on? 

Thanks,
Ryan

**Comment by akash07k** [comment]
Hi Ryan,
Thanks for the newer build,
Unfortunately, I too am still unable to get it working. The problem remains the same regarding the caching stuff.

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23, @akash07k,

Thank you for trying. I think as a next step, I might come up with a more 'standalone' test to check caching mechanism in isolation (i.e. a very simple test.exe). I want to make sure it's indeed a bug purely within OpenVINO itself, as opposed to something that the Audacity plugin is doing wrong which could be triggering such behavior. 

At the very least, I'm thinking about adding the ability to disable caching for the next official drop. Either through some kind environment variable (this is sort of ugly). Or, I can envision this kind of setting sitting nicely as a 'plugin-level' setting (as opposed to exposing it as a check-box for each individual feature).

Thanks,
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23, @akash07k,

Hope you're both doing well. There is a new OpenVINO release that may resolve some caching bugs. I've placed an updated zip file for you to try (if you're still willing to help check), here:  [Audacity-OpenVINO-2023.3-Module-3.4.2-R1.zip](https://drive.google.com/file/d/1v3PY0HyeAQ0AaGtRcCML8fXwdABImwd_/view?usp=drive_link)

Thanks in advance!
Ryan

**Comment by akash07k** [comment]
Thanks @RyanMetcalfeInt8 ,
I'll try this new  build and will let you know the results soon.

> Hi @Rosenheim23, @akash07k,
> 
> Hope you're both doing well. There is a new OpenVINO release that may resolve some caching bugs. I've placed an updated zip file for you to try (if you're still willing to help check), here: [Audacity-OpenVINO-2023.3-Module-3.4.2-R1.zip](https://drive.google.com/file/d/1v3PY0HyeAQ0AaGtRcCML8fXwdABImwd_/view?usp=drive_link)
> 
> Thanks in advance! Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Thanks for following up so quickly. I will let OpenVINO team know that the issue is still observed with latest (2023.3) release.

Regards,
Ryan

**Comment by teoskoki** [comment]
> Hi Christian,
> 
> Thanks for following up so quickly. I will let OpenVINO team know that the issue is still observed with latest (2023.3) release.
> 
> Regards, Ryan

Dear Ryan,
Thank you so much that you've done so much to solve this problem.
I'm writing here to tell you that I have THE SAME step by step issue. Totally the same. In 4 words - "It works only once".

And I've done everything that you've suggested. Including your last sulution. And it helped only to reset the problem. After I installed your updates, the separation worked only once, and that's it.

Could you please let me know if the team (and you) already found the solution.

PS: I have Intel Core i5-7300U CPU Windows 7 Professional sp1 x64

Thanks in advance

**Comment by teoskoki** [comment]
In your #68 I've found an additional solution - to remove %appdata%\audacity\openvino-model-cache folder before running music separation - it really works. If I delete the folder each time before using Audacity. At least it works! So, I'm glad. And thanks again.

If there is (or could be as a result of this unsolved problem) a plugin which deletes this folder each new time automatically, it would be great.

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23, @teoskoki, @akash07k,

Hope all is well with you. I just want to let you know that there is an updated version of these plugins available here: https://github.com/intel/openvino-plugins-ai-audacity/releases/tag/v3.5.0-R2

I wish I could say that I am confident that this 'cache' issue is resolved, but I *still* haven't been able to reproduce it on any of my running systems since it was first reported here. There have been plenty of other users that have hit it though, unfortunately.

Since we have upgraded to OpenVINO 2024.0, there's a chance that there was some bug fix included there that will resolve this issue... it would be very helpful to us if someone could let us know whether or not the same issue is present in this updated release.

Thanks very much!
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Thank you for giving it a shot. Hmm, I'll have to figure out how to get to the bottom of that.

Thanks again,
Ryan

**Comment by Helmut002** [comment]
Hi Ryan, we know each other from my problem with the openCL.dll that You solved in a very great way.
I have the same problem with the openvino-model-cache folder. Before I start an openvino-session I must delete the blob-file. It is not necessary to delet the whole folder - only the blob must be deleted. It is also not necessary to close aud restart audacity.
This problem exists on my both machines. One older PC with win10 aun one new PC with win11.

**Comment by RyanMetcalfeInt8** [comment]
Hi @Helmut002,

Sorry to hear that you also are experiencing this issue. Just curious, are you running the German version of Windows as well? I know that Christian, who reported this issue, is using the German version.

Thanks,
Ryan

**Comment by Helmut002** [comment]
Hi Ryan,
in both pc's I use the german version.
[GerätenamePCW11.docx](https://github.com/intel/openvino-plugins-ai-audacity/files/15145729/GeratenamePCW11.docx)
[GerätenameHelmut.docx](https://github.com/intel/openvino-plugins-ai-audacity/files/15145733/GeratenameHelmut.docx)
And I use audacity 3.5.1 with the right version of openvino (audacity-win-v3.5.1-R2-64bit-OpenVINO-AI-Plugins)

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23, @Helmut002,

I am finally able to reproduce this issue. It turns out, that if I simply select ```Deutsch``` as the language to install,
![image](https://github.com/intel/openvino-plugins-ai-audacity/assets/107415876/f2bf1539-55f7-4e4b-bc84-f5a59eac08d5)

then I am able to reproduce the issue, exactly as you (and others) have described. On the same system, if I uninstall Audacity, remove the cached blobs, and re-install in English -- then everything works fine. It's very, very strange.

Furthermore, if I have ```English``` installation, run Music Separation (to produce the cached blob), and then uninstall & install ```Deutsch``` version -- then Music Separation consistently runs okay (as it now is picking up the blobs saved from English installation).

I'll dig into this more..

**Comment by RyanMetcalfeInt8** [comment]
Hi @Rosenheim23,

Ah, good to know! I narrowed it down to a locale issue -- Looks like the OpenVINO caching mechanism is sensitive to the numeric locale that is set. I'll submit this issue to the OpenVINO project, and in the meantime I'll implement a workaround for this to be included into the next patch release.

Looking forward to finally getting to resolve this one!

Thanks!
Ryan

**Comment by RyanMetcalfeInt8** [comment]
Hi Christian,

Thank you for confirming that it's all working now with the latest release! I had intended to add a comment here to let you know that the new release was available... I must have forgot to do that, so I apologize..

Anyway, thanks for all of your testing (and patience) -- I think we can finally close this issue!

Best Regards,
Ryan
